src/sentry/filters/preprocess_hashes.py

```diff
@@ -1,13 +1,74 @@
 from __future__ import absolute_import
 
+import functools
+import six
+
 from django.core.cache import cache, get_cache, InvalidCacheBackendError
 
+from sentry.interfaces.base import get_interfaces
+from sentry.interfaces.exception import Exception as ExceptionInterface, SingleException
+from sentry.interfaces.stacktrace import Frame, Stacktrace
+from sentry.event_manager import _get_hashes_from_fingerprint, md5_from_hash
+
 
 try:
     hash_cache = get_cache('preprocess_hash')
 except InvalidCacheBackendError:
     hash_cache = cache
 
 
+class UnableToGenerateHash(Exception):
+    pass
+
+
 def get_raw_cache_key(project_id, event_id):
     return 'e:raw:{1}:{0}'.format(project_id, event_id)
+
+
+def get_preprocess_hash_inputs(event):
+    return get_preprocess_hash_inputs_with_reason(event)[1]
+
+
+def get_preprocess_hash_inputs_with_reason(data):
+    interfaces = get_interfaces(data)
+    platform = data['platform']
+    for interface in six.itervalues(interfaces):
+        kwargs = {'is_processed_data': False}
+        if isinstance(interface, SingleException):
+            kwargs['platform'] = platform
+        elif isinstance(interface, ExceptionInterface) or isinstance(interface, Stacktrace) or \
```

Comment:
You can do `isinstance(interface, (ExceptionInterface, Stacktrace, Frame))` here

Reply:
ooo